sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusTargetScaler.cs

```diff
@@ -0,0 +1,94 @@
+using Microsoft.Azure.WebJobs.ServiceBus.Listeners;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Globalization;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.WebJobs.Extensions.ServiceBus.Listeners
+{
+    public class ServiceBusTargetScaler : ITargetScaler
+    {
+        private readonly string _functionId;
+        private readonly ServiceBusMetricsProvider _serviceBusMetricsProvider;
+        private readonly ServiceBusOptions _options;
+        private readonly bool _isSessionsEnabled;
+        private readonly bool _singleDispatch;
+        private readonly TargetScalerDescriptor _targetScalerDescriptor;
+        private readonly string _entityPath;
+        private readonly ILogger _logger;
+
+        public ServiceBusTargetScaler(
+            string functionId,
+            string entityPath,
+            ServiceBusEntityType entityType,
+            Lazy<ServiceBusReceiver> receiver,
+            Lazy<ServiceBusAdministrationClient> administrationClient,
+            ServiceBusOptions options,
+            bool isSessionsEnabled,
+            bool singleDispatch,
+            ILoggerFactory loggerFactory
+            )
+        {
+            _functionId = functionId;
+            _serviceBusMetricsProvider = new ServiceBusMetricsProvider(entityPath, entityType, receiver, administrationClient, loggerFactory);
+            _entityPath = entityPath;
+            _targetScalerDescriptor = new TargetScalerDescriptor(functionId);
+            _logger = loggerFactory.CreateLogger<ServiceBusTargetScaler>();
+            _options = options;
+            _singleDispatch = singleDispatch;
+            _isSessionsEnabled = isSessionsEnabled;
+        }
+
+        public TargetScalerDescriptor TargetScalerDescriptor => _targetScalerDescriptor;
+
+        public async Task<TargetScalerResult> GetScaleResultAsync(TargetScalerContext context)
+        {
+            long activeMessageCount = await _serviceBusMetricsProvider.GetMessageCount().ConfigureAwait(false);
+            return GetScaleResultInternal(context, activeMessageCount);
+        }
+
+        internal TargetScalerResult GetScaleResultInternal(TargetScalerContext context, long messageCount)
+        {
+            int concurrency;
+
+            if (!context.InstanceConcurrency.HasValue)
+            {
+                if (!_singleDispatch)
+                {
+                    concurrency = _options.MaxMessageBatchSize;
+                }
+                else
+                {
+                    if (_isSessionsEnabled)
+                    {
+                        concurrency = _options.MaxConcurrentSessions;
+                    }
+                    else
+                    {
+                        concurrency = _options.MaxConcurrentCalls;
+                    }
+                }
+            }
+            else
+            {
+                concurrency = context.InstanceConcurrency.Value;
+            }
+
```

Comment:
We should validate that whatever concurrency value we end up using is > 0 (to ensure we don't do division by 0). Suggestion: throw ArgumentOutofRangeException?